src/user.rs

```diff
@@ -1,50 +1,54 @@
-use matrix_sdk::instant::Instant;
 use matrix_sdk::room::Room;
+use matrix_sdk::ruma::api::client::error::ErrorKind;
 use matrix_sdk::ruma::api::client::r0::account::register;
 use matrix_sdk::ruma::api::client::r0::room::create_room;
-use matrix_sdk::ruma::api::client::r0::uiaa::{AuthData, Dummy};
+use matrix_sdk::ruma::api::client::r0::uiaa::{AuthData, Dummy, UiaaResponse};
+use matrix_sdk::ruma::api::error::{FromHttpResponseError::*, ServerError::*};
 use matrix_sdk::ruma::events::room::message::MessageEventContent;
 use matrix_sdk::ruma::events::{AnyMessageEventContent, SyncMessageEvent};
+use matrix_sdk::HttpError::UiaaError;
 use matrix_sdk::{Client, ClientConfig, RequestConfig, SyncSettings};
 use rand::Rng;
 use serde::Serialize;
 use std::sync::Arc;
-use std::time::Duration;
+use std::time::{Duration, Instant};
+use tokio::sync::mpsc::Sender;
 use tokio::sync::Mutex;
 
 use matrix_sdk::ruma::{assign, RoomId, UserId};
 
-use crate::metrics::Metrics;
+use crate::events::Event;
 use crate::text::get_random_string;
 
 const PASSWORD: &str = "asdfasdf";
 
-pub struct Disconnected {
-    metrics: Metrics,
-}
-pub struct Registered {
-    metrics: Metrics,
-}
-pub struct LoggedIn {
-    metrics: Metrics,
-}
-#[derive(Default, Clone)]
+pub struct Disconnected;
+pub struct Registered;
+pub struct LoggedIn;
+#[derive(Clone)]
 pub struct Synching {
     rooms: Arc<Mutex<Vec<RoomId>>>,
-    metrics: Metrics,
 }
 
 #[derive(Clone)]
 pub struct User<State> {
     id: UserId,
     client: Arc<tokio::sync::Mutex<Client>>,
+    tx: Sender<Event>,
```

Comment:
nit: For a higher level abstraction I would use a type alias to hide the `mpsc` detail or a new type altogether to keep the messages used before for reporting.